src/util/taproot.rs

```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    /// the output key point as Q = P + int(hashTapTweak(bytes(P)))G.
+    /// See also [`TaprootSpendInfo::tap_tweak`].
+    /// Refer to BIP 341 footnote (Why should the output key always have
+    /// a taproot commitment, even if there is no script path?) for more details
+    ///
+    pub fn new_key_spend<C: secp256k1::Verification>(
+        secp: &Secp256k1<C>,
+        internal_key: schnorr::PublicKey,
+        merkle_root: Option<sha256::Hash>,
+    ) -> Self {
+        let tweak = compute_tap_tweak(merkle_root, internal_key);
+        let mut output_key = internal_key;
+        // # Panics:
+        //
+        // This would return Err if the merkle root hash is the negation of the secret
+        // key corresponding to the internal key.
+        // Because the tweak is derived as specified in BIP341 (hash output of a function),
+        // this is unlikely to occur (1/2^128) in real life usage, it is safe to unwrap this
+        let parity = output_key
+            .tweak_add_assign(&secp, &tweak)
+            .expect("Cryptographically unreachable code");
+        Self {
+            internal_key: internal_key,
+            merkle_root: merkle_root,
+            output_key_parity: parity,
+            output_key: output_key,
+            script_map: BTreeMap::new(),
+        }
+    }
+
+    /// Obtain the tweak and parity used to compute the output_key
+    pub fn tap_tweak(&self) -> sha256::Hash {
+        compute_tap_tweak(self.merkle_root, self.internal_key)
+    }
+
+    /// Output key(the key used in script pubkey) from Spend data. See also
+    /// [`TaprootSpendInfo::output_key_parity`]
+    pub fn output_key(&self) -> schnorr::PublicKey {
+        self.output_key
+    }
+
+    /// Parity of the output key. See also [`TaprootSpendInfo::output_key`]
+    pub fn output_key_parity(&self) -> bool {
+        self.output_key_parity
+    }
+
+    // Internal function to compute [`TaprootSpendInfo`] from NodeInfo
+    fn from_node_info<C: secp256k1::Verification>(
+        secp: &Secp256k1<C>,
+        internal_key: schnorr::PublicKey,
+        node: NodeInfo,
+    ) -> TaprootSpendInfo {
+        // Create as if it is a key spend path with the given merkle root
+        let mut info = TaprootSpendInfo::new_key_spend(secp, internal_key, Some(node.hash));
+        for leaves in node.leaves {
+            let key = (leaves.script, leaves.ver);
+            let value = leaves.merkle_branch;
+            match info.script_map.get_mut(&key) {
+                Some(set) => {
+                    set.insert(value);
+                    continue; // NLL fix
+                }
+                None => {}
+            }
+            let mut set = BTreeSet::new();
+            set.insert(value);
+            info.script_map.insert(key, set);
+        }
+        info
+    }
+
+    /// Access the internal script map
+    pub fn as_script_map(&self) -> &BTreeMap<(Script, LeafVersion), BTreeSet<TaprootMerkleBranch>> {
```

Comment:
`BTreeMap<(Script, LeafVersion), BTreeSet<TaprootMerkleBranch>>` looks like a good candidate for at least type alias – or newtype with its own API, taking part of the code redundancy load from `TaprootSpendInfo`